stackslib/src/net/stackerdb/config.rs

```diff
@@ -123,108 +138,250 @@ impl StackerDBConfig {
-                    "Contract {} stipulated more than maximum number of slots for one signer ({})",
-                    contract_id, STACKERDB_INV_MAX
+                    "StackerDB fn `{contract_id}.{STACKERDB_PAGE_COUNT_FUNCTION}` returned page count outside of u32 range",
                 );
-                warn!("{}", &reason);
+                warn!("{reason}");
+                NetError::InvalidStackerDBContract(
+                    contract_id.clone(),
+                    reason,
+                )
+            }
+        )
+    }
+
+    fn parse_slot_entry(
+        entry: ClarityValue,
+        contract_id: &QualifiedContractIdentifier,
+    ) -> Result<(StacksAddress, u32), String> {
+        let ClarityValue::Tuple(slot_data) = entry else {
+            let reason = format!(
+                "StackerDB fn `{contract_id}.{STACKERDB_SLOTS_FUNCTION}` returned non-tuple slot entry",
+            );
+            return Err(reason);
+        };
+
+        let Ok(ClarityValue::Principal(signer_principal)) = slot_data.get("signer") else {
+            let reason = format!(
+                "StackerDB fn `{contract_id}.{STACKERDB_SLOTS_FUNCTION}` returned tuple without `signer` entry of type `principal`",
+            );
+            return Err(reason);
+        };
+
+        let Ok(ClarityValue::UInt(num_slots)) = slot_data.get("num-slots") else {
+            let reason = format!(
+                "StackerDB fn `{contract_id}.{STACKERDB_SLOTS_FUNCTION}` returned tuple without `num-slots` entry of type `uint`",
+            );
+            return Err(reason);
+        };
+
+        let num_slots = u32::try_from(*num_slots)
+            .map_err(|_| format!("Contract `{contract_id}` set too many slots for one signer (max = {STACKERDB_PAGE_MAX})"))?;
+        if num_slots > STACKERDB_PAGE_MAX {
+            return Err(format!("Contract `{contract_id}` set too many slots for one signer (max = {STACKERDB_PAGE_MAX})"));
+        }
+
+        let PrincipalData::Standard(standard_principal) = signer_principal else {
+            return Err(format!(
+                "StackerDB contract `{contract_id}` set a contract principal as a writer, which is not supported"
+            ));
+        };
+        let addr = StacksAddress::from(standard_principal.clone());
+        Ok((addr, num_slots))
+    }
+
+    fn eval_signer_slots(
+        chainstate: &mut StacksChainState,
+        burn_dbconn: &dyn BurnStateDB,
+        contract_id: &QualifiedContractIdentifier,
+        tip: &StacksBlockId,
+    ) -> Result<Vec<(StacksAddress, u32)>, NetError> {
+        let page_count = Self::eval_page_count(chainstate, burn_dbconn, contract_id, tip)?;
+        if page_count == 0 {
+            debug!("StackerDB contract {contract_id} specified zero pages");
+            return Ok(vec![]);
+        }
+        let mut return_set: Option<Vec<(StacksAddress, u32)>> = None;
+        let mut total_num_slots = 0u32;
+        for page in 0..page_count {
```

Comment:
I think it might make sense to have a hard bound on the number of pages allowed, so this doesn't turn into a DoS vector in other StackerDB instances.  What about 4096?  Then, you can have ~2^16 StackerDB slots, and if each of those is 16 MB, that would allow the creation of a StackerDB of up to ~250 TB.

Reply:
Related to my comment above about the "clarity list size" versus the "page size" -- if we enforce the clarity list size constraint, I'd be inclined to just set the page count limit at `2` for now. If there's another use case for expanding that, it could always be addressed later.